src/ucp/api/ucp.h

```diff
@@ -2049,6 +2049,76 @@ ucs_status_ptr_t ucp_tag_send_nb(ucp_ep_h ep, const void *buffer, size_t count,
                                  ucp_datatype_t datatype, ucp_tag_t tag,
                                  ucp_send_callback_t cb);
 
+/**
+ * @ingroup UCP_COMM
+ * @brief Non-blocking tagged-send operations with user provided request
+ *
+ * This routine sends a messages that is described by the local address @a
+ * buffer, size @a count, and @a datatype object to the destination endpoint
+ * @a ep. Each message is associated with a @a tag value that is used for
+ * message matching on the @ref ucp_tag_recv_nbr "receiver".
+ *
+ * The routine is non-blocking and therefore returns immediately, however
+ * the actual send operation may be delayed. The send operation is considered
+ * completed when it is safe to reuse the source @e buffer. If the send
+ * operation is completed immediately the routine returns UCS_OK.
+ *
+ * If the operation is @b not completed immediately and no error reported
+ * then the UCP library will fill a user provided @a req and
+ * return UCS_INPROGRESS status. In order to monitor completion of the
+ * operation @ref ucp_request_check_status() should be used.
+ *
+ * The @a req filled by this routine is going to be completed faster than
+ * the one returned by @ref ucp_tag_send_nb(). To achieve this the routine
+ * uses bcopy and a higher rendezvous threshold. The threshold is controlled
+ * by the UCX_SEND_NBR_RNDV_THRESH environment variable.
+ *
+ * @note This routine provides a convinient and efficient way to implement
+ *       a blocking send pattern. For example MPI_send() pseudo code may
+ *       look like this:
+ *
+ * @code
+ * MPI_send(...)
+ * {
+ *     char request[ucp_request_size];
+ *     ucs_status_t status;
+ *
+ *     status = ucp_tag_send_nbr(ep, ..., request + ucp_request_size);
+ *     if (status != UCS_INPROGRESS) {
+ *         return status;
+ *     }
+ *
+ *     do {
+ *         ucp_worker_progress(worker);
+ *         status = ucp_request_check_status(request + ucp_request_size);
```

Comment:
This trick is based on internal implementation, looks dangerous.

Reply:
We already do things this way in ucp_tag_recv_nbr()

Reply:
Hmm.. I don't see this hack in ucp_tag_recv_nbr() docs.

Reply:
Maybe better to add smth like this to API?
```
ucs_status ucp_nbr_check_status(void *user_request)
{
    return ucp_request_check_status(user_request + ucp_request_size);
}
```

Reply:
from ucp_tag_recv_nbr() doc:

>Request handle allocated by the user. There should
> be at least UCP request size bytes of available
 >space before the @a req. The size of UCP request
 > can be obtained by @ref ucp_context_query function.

and
> In order to monitor completion of the operation
 > @ref ucp_request_check_status or @ref ucp_tag_recv_request_test should be
 > used.

Reply:
@evgeny-leksikov user still has to know UCP request size in order to allocate enough room for it

Reply:
but there is nothing about offset

Reply:
what do you mean? no documentation?

Reply:
@yosefe yes, but it's very strange to pass pointer to end of allocated memory...

Reply:
yes, following to current docs i would expect to pass pointer to allocated request to check_status_func but not to its end

Reply:
@evgeny-leksikov ucp_tag_recv_nbr() clearly states that you need to pass a pointers to an end. Requests allocated internally by ucp already return such a pointer to a caller. ex. ucp_tag_send_nb()

Reply:
i see.. thanks for clarifications! but i still dislike this magic :(